File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int left=0,right=n-1;
        while(left<=right){
            int mid = left + (right-left)/2;
            if(nums[mid]==target){
                return mid;
            }
            if(nums[left] <= nums[mid]){
                // if left half is sorted
                if(nums[left]<=target && nums[mid]>=target){
                   // if in range of sorted part 
                   right = mid;
                }else{
                    left = mid+1;
                }
            }else{
                // if right half is sorted
                if(nums[mid]<=target && nums[right]>=target){
                   // if in range of sorted part 
                   left = mid;
                }else{
                    right = mid-1;
                }
            }


        }
        return -1;
    }
};

File: Data Structures & Algorithms/search-insert-position/submission-1.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        /*

        */
        int lb=0;
        int ub=nums.length-1;
        int mid = 0;
        while(lb<ub){
            mid = (lb+ub)/2;
            if(nums[mid]==target){
                return mid;
            }else if(nums[mid]<target){
                lb = mid+1;
            }else{
                ub = mid-1;
            }
        }
        System.out.println(mid+"| ub: "+ub+" | lb:"+lb);
        if(ub==-1)return 0;
        if(ub==nums.length)return ub-1;
        if(nums[ub]<target){
            return ub+1;
        }else{
            return ub;
        }
    }
}